anchor_optimization/optimize_anchors.py

```diff
@@ -0,0 +1,347 @@
+
+    if ratios % 2 != 1:
+        raise Exception('The number of ratios has to be odd.')
+
+    entries = np.zeros((0, 4))
+    max_x = 0
+    max_y = 0
+
+    updating = 'immediate'
+    if threads > 1:
+        # when the number of threads is > 1 then 'updating' is set to
+        # 'deferred' by default (see the documentation of
+        # "scipy.optimize.differential_evolution())
+        updating = 'deferred'
+
+    if seed is None:
+        seed = np.random.RandomState()
+    else:
+        seed = np.random.RandomState(seed)
+
+    if verbose:
+        print('Loading object dimensions.')
+
+    with _open_for_csv(annotations) as file:
+        for line, row in enumerate(csv.reader(file, delimiter=',')):
+            x1, y1, x2, y2 = list(map(lambda x: int(x), row[1:5]))
+
+            if not x1 or not y1 or not x2 or not y2:
+                continue
+
+            if resize:
+                # Concat base path from annotations file follow retinanet
+                base_dir = os.path.split(annotations)[0]
+                relative_path = row[0]
+                image_path = os.path.join(base_dir, relative_path)
+                img = Image.open(image_path)
+
+                if hasattr(img, "shape"):
+                    image_shape = img.shape
+                else:
+                    image_shape = (img.size[0], img.size[1], 3)
+
+                scale = compute_resize_scale(
+                    image_shape, min_side=image_min_side, max_side=image_max_side)
+                x1, y1, x2, y2 = list(map(lambda x: int(x) * scale, row[1:5]))
+
+            max_x = max(x2, max_x)
+            max_y = max(y2, max_y)
+
+            if include_stride:
+                entry = np.expand_dims(np.array([x1, y1, x2, y2]), axis=0)
+                entries = np.append(entries, entry, axis=0)
+            else:
+                width = x2 - x1
+                height = y2 - y1
+                entry = np.expand_dims(
+                    np.array([-width / 2, -height / 2, width / 2, height / 2]), axis=0)
+                entries = np.append(entries, entry, axis=0)
+
+    image_shape = [max_y, max_x]
+
+    if verbose:
+        print('Optimising anchors.')
+
+    bounds = []
+
+    for i in range(int((ratios - 1) / 2)):
+        bounds.append((1, 4))
+
+    for i in range(scales):
+        bounds.append((0.4, 2))
+
+    update_state = None
+    if threads == 1:
+        update_state = state
+
+    ARGS = (entries,
+            image_shape,
+            objective,
+            ratios,
+            include_stride,
+            SIZES,
+            STRIDES,
+            verbose,
+            update_state,
+            # return a single value ('to_tuple' parameter is set to False)
+            False,
+            threads)
+
+    result = scipy.optimize.differential_evolution(func=average_overlap,
+                                                   args=ARGS,
+                                                   # pass the *args as a tuple
```

Comment:
Put the comment into a single line and move it above the args parameter.

Reply:
thanks, I missed this one.